explorer/interpreter/type_checker.cpp

```diff
@@ -1231,33 +1274,42 @@ class TypeChecker::ConstraintTypeBuilder {
     return Success();
   }
 
-  class ImplsInScopeTracker {
+  class ConstraintsInScopeTracker {
     friend class ConstraintTypeBuilder;
 
    private:
-    int num_added = 0;
+    int num_impls_added = 0;
+    int num_equals_added = 0;
   };
 
-  // Brings all the `impl`s accumulated so far into the given impl scope.
-  // If this will be called more than once, an ImplsInScopeTracker can be
-  // provided to avoid adding the same impls more than once.
-  void BringImplsIntoScope(
-      const TypeChecker& type_checker, Nonnull<ImplScope*> impl_scope,
-      std::optional<Nonnull<ImplsInScopeTracker*>> tracker = std::nullopt) {
+  // Brings all the constraints accumulated so far into the given impl scope,
+  // as if we built the constraint type and then added it into the scope. If
+  // this will be called more than once, an ImplsInScopeTracker can be provided
+  // to avoid adding the same impls more than once.
+  void BringConstraintsIntoScope(const TypeChecker& type_checker,
+                                 Nonnull<ImplScope*> impl_scope,
+                                 Nonnull<ConstraintsInScopeTracker*> tracker) {
+    // Figure out which constraints we're going to add.
+    int first_impl_to_add =
+        std::exchange(tracker->num_impls_added, impl_constraints_.size());
+    int first_equal_to_add =
+        std::exchange(tracker->num_equals_added, equality_constraints_.size());
     llvm::ArrayRef<ConstraintType::ImplConstraint> impl_constraints =
         impl_constraints_;
-    if (tracker) {
-      impl_constraints = impl_constraints.drop_front((*tracker)->num_added);
-      (*tracker)->num_added = impl_constraints_.size();
+
+    // Add all of the new constraints.
+    impl_scope->Add(impl_constraints.drop_front(first_impl_to_add), llvm::None,
+                    llvm::None, GetSelfWitness(), type_checker);
+    for (int i = first_equal_to_add;
+         i != static_cast<int>(equality_constraints_.size()); ++i) {
+      impl_scope->AddEqualityConstraint(
+          arena_->New<EqualityConstraint>(equality_constraints_[i]));
```

Comment:
```suggestion
    for (auto& constraint : llvm::ArrayRef(equality_constraints_).drop_front(first_equal_to_add)) {
      impl_scope->AddEqualityConstraint(arena_->New<EqualityConstraint>(constraint));
```

Not sure this is better, but the static_cast had me thinking and I figured I might suggest the idea.

Reply:
Done and refactored a little to make the two cases look more similar to each other.